ENH: Modify intersect1d to return common indices

cagostino opened the pull request:
Proposed this idea in previous pull request (https://github.com/numpy/numpy/pull/10683) and made suggested changes for implementing this idea to have relevant common indices that correspond to the values in the intersection of the two arrays.

cagostino committed:
added function commonpts1d

cagostino committed:
Update arraysetops.py

cagostino committed:
Update arraysetops.py

cagostino committed:
Merge pull request #1 from cagostino/cagostino-patch-3

update intersect1d to have common points

cagostino committed:
modified intersect1d to return common indices

Proposed this idea in previous pull request (https://github.com/numpy/numpy/pull/10683) and made suggested changes for implementing this idea to have relevant common indices that correspond to the values in the intersection of the two arrays.

cagostino committed:
Merge pull request #2 from cagostino/cagostino-patch-1

modified intersect1d for returning common indices

jaimefrio commented on numpy/lib/arraysetops.py:
```diff
@@ -332,14 +338,26 @@ def intersect1d(ar1, ar2, assume_unique=False):
+    if return_indices:
+      a_inds, = np.in1d(ar1,int1d).nonzero()
+      b_inds, = np.in1d(ar2,int1d).nonzero()
+      return int1d,a_inds,b_inds
```
There is no `a` and `b` referenced elsewhere in this function, better call these `inds1` and `inds2`

jaimefrio replied:
Or, if you ask me, `indices1` and `indices2`...

jaimefrio commented:
I don't like the fact that, as implemented, the indices are sorted differently than the common values ,e.g.:

```
>>> np.intersect1d([5, 3, 1, 4, 9, 11], [2, 5, 8, 4, 12, 10, 3], return_indices=True)
array([3, 4, 5]), array([0, 1, 3]), array([1, 3, 6])
```

I think it would be much better if the return was:

```
array([3, 4, 5]), array([1, 3, 0]), array([6, 3, 1])
```

so that `array1[indices1] == intersection` and `array2[indices2] == intersection`.

Something like this would do the trick if `assume_unique = True`:

```
aux = np.concatenate([a, b])
aux_sort_indices = np.argsort(c, kind='mergesort')
aux_sorted = aux[aux_sort_indices]
mask = aux_sorted[1:] == aux_sorted[:-1]
int1d = aux_sorted[:-1][mask]
ar1_indices = aux_sort_indices[:-1][mask]
ar2_indices = aux_sort_indices[1:][mask] - ar1.size
```

We would only want to do sorting like this if the indices are requested, as it is slower than the current `aux.sort`, and the code assumes that both `ar1` and `ar2` are indeed 1D arrays, so we should make sure they actually are before e.g. calling `.size` on `ar1`. And handling `assume_unique=False` will require some extra work on the indices that needs to be figured out.

jaimefrio commented:
Also, a new parameter is a change to the API, so it has to be shared with the mailing list for discussion. I think this is a nice addition, but we want to allow everyone to have a say: could you write an e-mail linking here?

cagostino commented:
Yes that was the original intention with regard to the indices being sorted. I originally had it written with a for loop to maintain the structure between them but isin (which I'm not as familiar with)  was suggested to vectorize it. I'll try and make these changes (probably not for a few days because traveling) and write up an email. Thank you so much for the suggestions!

charris commented:
Enhancements need a mention in the `doc/release/1.15.0-notes.rst`.

cagostino committed:
update intersect1d with suggested changes

cagostino commented:
I made most of the suggested changes above but I'm curious if anyone has a good idea for what might be the best vectorized way to retain the indices for the two respective arrays. 
This is how I originally had it when I suggested it be a separate function from intersect1d. I'll send an email to the mailing list looking for suggestions as well.

`  
    first = []
    second =[]
    commvals = np.intersect1d(arr1,arr2)
    for i in range(len(commvals)):
        val1 = np.where(arr1 == commvals[i])[0][0]
        val2 = np.where(arr2 == commvals[i])[0][0]
        first.append(val1)
        second.append(val2) #shouldn't happen
    comm1 =np.reshape(np.array(first),-1)
    comm2 = np.reshape(np.array(second),-1)
    return commvals,comm1,comm2
`

jaimefrio commented:
If `assume_unique=True`, I already proposed a vectorized approach above. If not, the answer depends a little on what we want to achieve. Say your two arrays are `[3, 4, 5, 3]` and `[2, 5, 3]`. The intersection is `[3, 5]`, and the indices for the 2nd array should be `[2, 1]`, but what do we want the indices for the first to be? `[0, 2]`? `[0, 3, 2]`? Not sure what the use case for this is supposed to be exactly, so I find it hard to tell.

cagostino commented:
AHHH I see what you mean. Yes that does seem like it could be problematic.
The specific use case which made me think of raising the issue is object ID numbers between two catalogs so the arrays themselves are unique sets to begin with. For the case you describe above. I think it would be best to have them be the indices of first occurrence, similar to how unique_indices in np.unique operates.

jaimefrio commented:
Yes, that seems to make sense, but it's an important point to highlight on the e-mail thread.

As for a vectorized solution for this when `assume_unique=False`, you would probably want to call `np.unique` with `return_indices=True`, store those indices, and instead of returning the indices of unique, use them to index into the stored indices to get your actual return array. It's confusing to write it as a parragraph, but if you think about it carefully it will hopefully make sense!

cagostino committed:
implemented return_indices options for intersect1d

I've tested out the above code and for the ~10 different test sets I've tried it's worked thus far. It's probably not the cleanest implementation but it works and is vectorized.

cagostino commented:
I did as you suggested and I'm pretty sure I've got a working implementation. Thank you for the tips and help!

eric-wieser commented on numpy/lib/arraysetops.py:
```diff
@@ -332,14 +338,45 @@ def intersect1d(ar1, ar2, assume_unique=False):
+    if return_indices:
+      if not assume_unique:
+        a1, ind1 = unique(ar1, return_index = True) #a1 not necessary
+        a2, ind2 = unique(ar2, return_index = True)
```
Strange formatting here - we always use multiples of four spaces to indent, and PEP8 says no spaces around keyword arguments

cagostino replied:
It's a relic i forgot to remove, I'll fix this.

eric-wieser commented on numpy/lib/arraysetops.py:
```diff
@@ -332,14 +338,45 @@ def intersect1d(ar1, ar2, assume_unique=False):
+        a2, ind2 = unique(ar2, return_index = True)
+        aux = np.concatenate([a1,a2])
+      else:
+        aux = np.concatenate([ar1, ar2])
```
Can these two branches be made the same? Ie, just re-use the name `ar1` to refer to the unique'd `ar1`, something like:

```
# make arrays unique if needed
if not assume_unique:
    if return_indices:
        ar1, ind1 = unique(ar1, return_index=True)
        ar2, ind2 = unique(ar2, return_index=True)
    else:
        ar1 = unique(ar1)
        ar2 = unique(ar2)
aux = np.concatenate([ar1, ar2])

# now special case
if return_indices:
   # the code you have below
```

jaimefrio reviewed (changes requested):
One other thing that worries me is that we are making an assumption that the array is 1D, but we are not enforcing it anywhere. This is not strictly a problme of your PR, but since we are redoing this function a lot we might as well try to fix it. If you are handed two n-D arrays I think you'll either get a cryptic error from concatenate, or if the dimensions match get an even more cryptic error further down or an entirely wrong result. I'm not sure what the best solution is, although it probably make sense to always ravel or flatten the inputs.

What we definitely need is some tests! Both for the new index returning functionality and for these weirder multidimensional corner cases.

There's still some work to do, but this is starting to shape up, thanks a lot!

cagostino committed:
cleaned up structure for intersect1d

cagostino commented:
I think that concern can be tackled by checking if the number of elements in the input array is also in the shape. I mention the shape in the case that someone has a 1-d array in a different orientation.  If the inputs are list rather than arrays then they could be hard casted as arrays to check for shape.
This would be like 
`
if type(ar1) == list:
    ar1 =np.copy(np.array(ar1))
if type(ar2) == list:
    ar2 = np.copy(np.array(ar2))
if ar1.size not in ar1.shape or ar2.size not in ar2.shape:
    raise DimensionError('Input array is not 1 dimensional')
`
or something similar

eric-wieser commented on numpy/lib/arraysetops.py:
```diff
@@ -332,14 +338,42 @@ def intersect1d(ar1, ar2, assume_unique=False):
+
+        return int1d, ar1_indices, ar2_indices
+
+    return int1d
```
This would make more sense in an `else`

eric-wieser commented on numpy/lib/arraysetops.py:
```diff
@@ -332,14 +338,42 @@ def intersect1d(ar1, ar2, assume_unique=False):
-        ar2 = unique(ar2)
+        if return_indices:
+            ar1, ind1 = unique(ar1, return_indices=True)
+            ar1, ind1 = unique(ar1, return_indices=True)
```
copy-paste error here

cagostino replied:
oops, got it thank you

cagostino committed:
fixed copy-paste error, added second test, changed a conditional

eric-wieser commented:
Looks like you didn't add any tests at all?

eric-wieser commented on numpy/lib/arraysetops.py:
```diff
@@ -343,12 +343,15 @@ def intersect1d(ar1, ar2, assume_unique=False, return_indices=False):
-    (array(1,2,4), array([0,1,3]), array([1,0,2]))
+    (array([1,2,4]), array([0,1,3]), array([1,0,2]))
+    >>> np.intersect1d([1,2,2,3,4,3,2],[1,8,4,2,2,3,2,3],return_indices=True)
+    (array([1, 2, 3, 4]), array([0, 1, 3, 4]), array([0, 3, 5, 2]))
```
These doctests are not run. Can you add some tests to the separate test files?

cagostino replied:
OH yes I wasn't aware of that. I'll add these there quickly.

cagostino committed:
Testing return_indices in intersect1d

cagostino committed:
Merge pull request #3 from cagostino/patch-2

Patch 2

cagostino committed:
formatting

eric-wieser commented on numpy/lib/tests/test_arraysetops.py:
```diff
@@ -34,7 +34,24 @@ def test_intersect1d(self):
+        assert_array_equal(c, ef)
+        assert_array_equal(a[i1], ef)
+        assert_array_equal(b[i2], ef)
     def test_setxor1d(self):
```
nit: should be a space before the start of this next test

eric-wieser commented on numpy/lib/tests/test_arraysetops.py:
```diff
@@ -34,7 +34,24 @@ def test_intersect1d(self):
         assert_array_equal([], intersect1d([], []))
-
+        
+        #return_indices
```
May as well add a `def test_intersect1d_indices(self)` here, which creates a new test rather than making the existing one longer

eric-wieser commented on numpy/lib/tests/test_arraysetops.py:
```diff
@@ -34,7 +34,24 @@ def test_intersect1d(self):
+        assert_array_equal(b[i2],ee)
+        #non-unique case
+        a = np.array([1,2,2,3,4,3,2])
+        b = np.array([1,8,4,2,2,3,2,3])
```
Should put spaces after commas everywhere in these tests

cagostino committed:
created separate test function  for intersect1d indices, added spaces after commas

eric-wieser commented on numpy/lib/arraysetops.py:
```diff
@@ -311,11 +311,17 @@ def intersect1d(ar1, ar2, assume_unique=False):
+    comm1 : ndarray
+        Indices of common points in first array.
+    comm2 : ndarray
+        Indices of common points in second array.
```
"elements", not "points", would match the other descriptions

eric-wieser commented on numpy/lib/tests/test_arraysetops.py:
```diff
@@ -34,7 +34,25 @@ def test_intersect1d(self):
 
         assert_array_equal([], intersect1d([], []))
-
+    def test_intersect1d_indices(self):
```
PEP8 needs a newline above this line, and below the last line of the function

cagostino replied:
okay i added one after the function as well (same problem)

eric-wieser commented on numpy/lib/arraysetops.py:
```diff
@@ -332,14 +338,45 @@ def intersect1d(ar1, ar2, assume_unique=False):
+    along with the intersected values:
+    
+    >>> np.intersect1d([1,2,3,4], [2,1,4,6], return_indices=True)
+    (array([1,2,4]), array([0,1,3]), array([1,0,2]))
```
Missing spaces around commas, especially in the output where what you have doesn't match what is printed.

I think this example would be more clear if made longer:
```python
>>> x = np.array([1,2,3,4])
>>> y = np.array([2,1,4,6])
>>> xy, x_ind, y_ind = np.intersect1d([1,2,3,4], [2,1,4,6], return_indices=True)
>>> xy
array([1, 2, 4])
>>> x_ind, y_ind
(array([0, 1, 3]), array([1, 0, 2]))
```

cagostino committed:
fixed up example and some style

cagostino committed:
fixed style

cagostino committed:
style change

eric-wieser commented on numpy/lib/arraysetops.py:
```diff
@@ -332,14 +338,49 @@ def intersect1d(ar1, ar2, assume_unique=False):
+    >>> xy, x[x_ind], y[y_ind]
+    (array([1, 2, 4]), array([1, 2, 4]), array([1, 2, 4]))
+    >>> np.intersect1d([1, 2, 2, 3, 4, 3, 2],[1, 8, 4, 2, 2, 3, 2, 3], return_indices=True)
+    (array([1, 2, 3, 4]), array([0, 1, 3, 4]), array([0, 3, 5, 2]))
```
I don't think this example is particularly useful or well explained - I'd be inclined to remove it

cagostino replied:
will do.

eric-wieser commented on numpy/lib/tests/test_arraysetops.py:
```diff
@@ -99,7 +119,7 @@ def assert_isin_equal(a, b):
         #array-likes as both arguments
         c = [(9, 8), (7, 6)]
-        d = (9, 7)
+        d = (9, 7) 
```
Not sure why this space was added, but would be better not to

cagostino replied:
uhh I'm not sure either. will fix.

cagostino committed:
removed one example

cagostino committed:
removed extra space

eric-wieser commented on numpy/lib/arraysetops.py:
```diff
@@ -311,11 +311,17 @@ def intersect1d(ar1, ar2, assume_unique=False):
-
+    return_indices : bool
+        If True, the indices which correspond to the interesction of the 
+        two arrays are returned. Default is False.
```
Needs `..versionadded:: 1.15.0`

cagostino committed:
added version number for return_indices

cagostino committed:
added 'return_indices' keyword for np.intersect1d

cagostino committed:
fixed formatting

eric-wieser commented on numpy/lib/arraysetops.py:
```diff
@@ -311,11 +311,20 @@ def intersect1d(ar1, ar2, assume_unique=False):
         can speed up the calculation.  Default is False.
-
+    return_indices : bool
+        If True, the indices which correspond to the interesction of the 
```
typo: should be `intersection`

eric-wieser commented on doc/release/1.15.0-notes.rst:
```diff
@@ -71,6 +71,14 @@ Creating a full iOS-compatible NumPy package requires building for the 5
+
+New keyword ``return_indices`` returns the indices of the two input arrays
+that correspond to the common elements. The first instance of a value is 
+used if there are multiple.
```
This last sentence should appear in the function docstring, and maybe isn't needed here

eric-wieser commented on numpy/lib/arraysetops.py:
```diff
@@ -311,11 +311,20 @@ def intersect1d(ar1, ar2, assume_unique=False):
     intersect1d : ndarray
         Sorted 1D array of common and unique elements.
+    comm1 : ndarray
+        Indices of common elements in first array.
```
Perhaps `arr1` instead of `first array`?

eric-wieser commented on doc/release/1.15.0-notes.rst:
```diff
@@ -71,6 +71,14 @@ Creating a full iOS-compatible NumPy package requires building for the 5
 combining these 5 compiled builds products into a single "fat" binary.
 
+``return_indices`` keyword added for ``np.intersect1d``
+-----------------------------------------
```
I think this needs the `-` characters to align

eric-wieser commented:
So close!

cagostino committed:
updated return_indices entry

cagostino committed:
fixed some typos and style

cagostino committed:
added bit about first instance of a value being used

eric-wieser committed:
Merge branch 'master' into master

eric-wieser committed:
STY: Fix comment formats

eric-wieser reviewed (dismissed):
Made some final tweaks and fixed the conflicts myself.

@jaimefrio, what's left to do here?

eric-wieser commented on numpy/lib/arraysetops.py:
```diff
@@ -332,14 +342,47 @@ def intersect1d(ar1, ar2, assume_unique=False):
+            ar2, ind2 = unique(ar2, return_index=True)
+        else:
+            ar1 = unique(ar1)
+            ar2 = unique(ar2)
```
I think you want an

```
else:
    # do the flattening that unique does
    ar1 = ar1.ravel()
    ar2 = ar2.ravel()
```

eric-wieser commented on numpy/lib/arraysetops.py:
```diff
@@ -311,11 +311,21 @@ def intersect1d(ar1, ar2, assume_unique=False):
     assume_unique : bool
```
On the line above, add "this will be flattened if it is not already 1D"

eric-wieser committed:
DOC: missing space

eric-wieser committed:
DOC: correct parameter names in docstring

eric-wieser commented on numpy/lib/arraysetops.py:
```diff
@@ -311,11 +311,21 @@ def intersect1d(ar1, ar2, assume_unique=False):
     intersect1d : ndarray
         Sorted 1D array of common and unique elements.
+    comm1 : ndarray
+        Indices of common elements in ar1. 
```
You should try to match the docs for `np.unique` here, and document this as

    The indices of the first occurrences of the common values in `arr1`. Only provided if `return_indices` is True.

cagostino committed:
made suggested changes

eric-wieser commented:
I think you misunderstood my comment above. I meant you need an `else` on the `if not assume_unique` case, to flatten the already-unique arrays. What you're doing now means that `unique` is not called at all when the array is _non_ unique

cagostino commented:
That makes more sense. I'll fix it, thanks.

cagostino committed:
fixed a mistake from previous update

also added documentation for comm1, comm2 to match doc from np.unique

eric-wieser commented on numpy/lib/arraysetops.py:
```diff
@@ -332,14 +345,49 @@ def intersect1d(ar1, ar2, assume_unique=False):
+            ar2 = unique(ar2)
+    else:
+        ar1 = ar1.ravel()
+        ar2 = ar2.ravel()
```
These lines need a test that hit them.

eric-wieser commented:
Maybe worth adding an example with a 2d case. In particular, you could show how to use `unravel_index` to reconstruct the 2d indices.

At any rate, you should definitely test a 2d case, both with and without `assume_unique`.

cagostino committed:
added in tests for 2d inputs

eric-wieser commented on numpy/lib/tests/test_arraysetops.py:
```diff
@@ -32,7 +32,46 @@ def test_intersect1d(self):
+        #non1d, unique inputs
+        a = np.array([[2,4,5,6],[7,8,2,15]])
+        b = np.array([[3,2,7,6],[10,12,8,9]])
+        c,i1,i2 = intersect1d(a,b,return_indices=True)
```
Did you mean to add `assume_unique` here?

Should be spaces around the commas too

eric-wieser commented on numpy/lib/tests/test_arraysetops.py:
```diff
@@ -32,7 +32,46 @@ def test_intersect1d(self):
+        #non1d, not assumed to be uniqueinputs
+        a = np.array([[2,4,5,6,6],[4,7,8,7,2]])
+        b = np.array([[3,2,7,7],[10,12,8,7]])
+        c,i1,i2 = intersect1d(a,b,return_indices=True)
```
Spaces around commas needed on these three lines

eric-wieser replied:
nevermind, fixed

eric-wieser committed:
STY: Add missing spaces around commas

eric-wieser commented on numpy/lib/tests/test_arraysetops.py:
```diff
@@ -32,7 +32,46 @@ def test_intersect1d(self):
+        #non1d, unique inputs
+        a = np.array([[2, 4, 5, 6], [7, 8, 2, 15]])
+        b = np.array([[3, 2, 7, 6], [10, 12, 8, 9]])
+        c, i1, i2 = intersect1d(a,b,return_indices=True)
```
Did you mean to add `assume_unique` here?

eric-wieser replied:
I've gone ahead and assumed that you did

eric-wieser committed:
Merge branch 'master' into master

eric-wieser committed:
TST: Correct array to actually be unique

eric-wieser committed:
STY: Spaces at beginning of comments

eric-wieser commented:
In the interest of getting this into 1.15, I've addressed all of my own nits. @jaimefrio, any final comments?

jaimefrio reviewed (approved):
It LGTM now. Thanks for all the work!
